### ISSUE TITLE
Query: Parameter Extractor should be eager at evaluating Convert nodes

### ISSUE THREAD

**Issue by smitpatel** [aspnet/EntityFrameworkCore#9361]
https://github.com/aspnet/EntityFramework/blob/dev/src/EFCore/Query/ExpressionVisitors/Internal/ParameterExtractingExpressionVisitor.cs#L278

While visiting UnaryExpression, we call base right away without checking if the current Expression is evaluabtable or not (like how VisitMember does it). So if there a closure variable wrapped inside a convert node, we would carry on the convert in Expression tree when we could have just computed the parameter after applying convert.
This can help issues like #9356 when the convert nodes are around a closure variables only.

**Comment by elepner** [comment]
Hello!
I was trying to approach this issue. [I have implemented eager evaluation of Convert](https://github.com/elepner/EntityFrameworkCore/commit/9c010c21a7c8cc17d16fd3587df4098dbe4fb447) nodes. It looked very simple, and now the engine pulls variable value if possible. However this fix broke tons of integration tests. They expect parameterized SQL like this:
```sql
@__TenantPrefix_0='B' (Size = 4000)
@__get_Item_0='ALFKI' (Size = 5)

SELECT TOP(1) [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE (([c].[CompanyName] LIKE @__TenantPrefix_0 + N'%' AND (LEFT([c].[CompanyName], LEN(@__TenantPrefix_0)) = @__TenantPrefix_0)) OR (@__TenantPrefix_0 = N'')) AND ([c].[CustomerID] = @__get_Item_0)
```
But after my fix it gets that:
```sql
@__TenantPrefix_0='B' (Size = 4000)

SELECT TOP(1) [c].[CustomerID], [c].[Address], [c].[City], [c].[CompanyName], [c].[ContactName], [c].[ContactTitle], [c].[Country], [c].[Fax], [c].[Phone], [c].[PostalCode], [c].[Region]
FROM [Customers] AS [c]
WHERE (([c].[CompanyName] LIKE @__TenantPrefix_0 + N'%' AND (LEFT([c].[CompanyName], LEN(@__TenantPrefix_0)) = @__TenantPrefix_0)) OR (@__TenantPrefix_0 = N'')) AND ([c].[CustomerID] = N'ALFKI')
```
They both produce same result, but string is different.
Is it possible to distinguish when when the method should try to cast the value and when try to rebuild input expression and transform it to ``Expression.Parameter``? Could you please help me and specify what I'm doing wrong?

**Comment by smitpatel** [comment]
@elepner - Thanks for taking time to implement this. I have left feedback on your commit which should get you closer to the final implementation. It would at least solve the issue with SQLs you have posted above. Feel free to post more question if you encounter any further roadblocks.

**Comment by SSkovboiSS** [comment]
@elepner are you still working on this?

I am currently fixing another issue related to `ParameterExtractingExpressionVisitor` and I would give this one a try as well.

**Comment by smitpatel** [comment]
This is fixed in 3.0 query pipeline. We only preserve convert node when they are required.
